### ISSUE TITLE
Handle event history with out-of-order eventIds.

### ISSUE THREAD

**Issue by kashimiz** [Azure/azure-functions-durable-js#38]
Discovered by reproducing #36. Durable Task Framework is not guaranteed to present history events with non-negative eventIds (e.g. TaskScheduled, TimerCreated) in order. From this code:

```javascript
const df = require("durable-functions");

module.exports = df.orchestrator(function* (context) {
  let helloTasks = [
    context.df.callActivity('SayHello', 'A'),
    context.df.callActivity('SayHello', 'B')
  ];
  yield context.df.Task.all(helloTasks);

  let moreHelloTasks = [
    context.df.callActivity('SayHello', 'C'),
    context.df.callActivity('SayHello', 'D')
  ];
  yield context.df.Task.all(moreHelloTasks);

  return 'OK';
});
```

the following event history is possible (abridged for clarity):

```
[
  // orchestrator orchestrator started, execution started
  { "EventId": 0, "EventType": "TaskScheduled", "Name": "SayHello" },  // A
  { "EventId": 1, "EventType": "TaskScheduled", "Name": "SayHello" },  // B
  // orchestrator completed, orchestrator started
  { "EventId": -1, "EventType": "TaskCompleted", "Result": ""Hello A!"", "TaskScheduledId": 0 },
  // orchestrator completed, orchestrator started
  { "EventId": -1, "EventType": "TaskCompleted", "Result": ""Hello B!"", "TaskScheduledId": 1 },
  { "EventId": 3, "EventType": "TaskScheduled", "Name": "SayHello" },  // D
  { "EventId": 2, "EventType": "TaskScheduled", "Name": "SayHello" },  // C
  // orchestrator completed, orchestrator started
  { "EventId": -1, "EventType": "TaskCompleted", "Result": ""Hello C!"", "TaskScheduledId": 2 },
  // orchestrator completed, orchestrator started
  { "EventId": -1, "EventType": "TaskCompleted", "Result": ""Hello C!"", "TaskScheduledId": 2 },
]
```

Tracing shows the shim incorrectly matches the `callActivity("SayHello", "D")` call to the history event returning ""Hello C!"", likely due to the (incorrect) assumption that the order in which tasks are scheduled === the order of TaskScheduled events in history. This can affect calls to `context.df.Task.any`.

**Comment by webuniverseio** [comment]
Hi, I feel I'm running into a similar problem with multiple Task.all and context.df.callActivityWithRetry. What is the status on this one? Do you know what is the problem? If yes, when fix will be released? It also mentions here that repro is needed - do you want me to provide a sample to reproduce this?

**Comment by ConnorMcMahon** [comment]
@webuniverseio,

The underlying cause for this incident was actually resolved in [this issue](https://github.com/Azure/azure-functions-durable-extension/issues/555) for the Durable Task extension. This was fixed in Durable Task extension 1.7.1.

If you are still seeing odd symptoms and you are on that version of the extension or greater, can you open an issue with some more details about your symptoms, ideally with a repro.
